File: CS/Form1.cs
using System;
using System.IO;
using System.Windows.Forms;
using System.Collections.Generic;
using DevExpress.XtraRichEdit;
using DevExpress.Portable.Input;

namespace RichEditNavigation {
    public partial class Form1 : Form {
        string basePath = "Documents";
        string startPage = "Index.html";
        List<string> historyItems = new List<string>();
        int currentHistoryItemIndex = 0;

        public Form1() {
            InitializeComponent();
            richEditControl1.ReadOnly = true;
            richEditControl1.Options.Hyperlinks.ModifierKeys = PortableKeys.None;
            NavigateTo(startPage, true);
        }

        private void richEditControl1_HyperlinkClick(object sender, HyperlinkClickEventArgs e) {
            BeginInvoke((MethodInvoker)delegate() {
                NavigateTo(e.Hyperlink.NavigateUri, true);
            });
            e.Handled = true;
        }

        private void richEditControl1_KeyDown(object sender, KeyEventArgs e) {
            if (currentHistoryItemIndex != 0 && e.KeyCode == Keys.Back)
                btnBackward_Click(btnBackward, EventArgs.Empty);
        }

        private void btnBackward_Click(object sender, System.EventArgs e) {
            NavigateTo(historyItems[--currentHistoryItemIndex], false);
        }

        private void btnForward_Click(object sender, System.EventArgs e) {
            NavigateTo(historyItems[++currentHistoryItemIndex], false);
        }

        private void NavigateTo(string page, bool addToHistory) {
            if (addToHistory) {
                while (historyItems.Count > currentHistoryItemIndex + 1)
                    historyItems.RemoveAt(historyItems.Count - 1);
                
                historyItems.Add(page);
                currentHistoryItemIndex = historyItems.Count - 1;
            }
            richEditControl1.LoadDocument(Path.Combine(basePath, page));
            UpdateUI();
        }

        private void UpdateUI() {
            btnBackward.Enabled = currentHistoryItemIndex != 0;
            btnForward.Enabled = currentHistoryItemIndex != historyItems.Count - 1;
        }
    }
}

File: VB/Form1.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.IO
Imports System.Windows.Forms
Imports System.Collections.Generic
Imports DevExpress.XtraRichEdit
Imports DevExpress.Portable.Input

Namespace RichEditNavigation
	Partial Public Class Form1
		Inherits Form
		Private basePath As String = "Documents"
		Private startPage As String = "Index.html"
		Private historyItems As New List(Of String)()
		Private currentHistoryItemIndex As Integer = 0

		Public Sub New()
			InitializeComponent()
			richEditControl1.ReadOnly = True
			richEditControl1.Options.Hyperlinks.ModifierKeys = PortableKeys.None
			NavigateTo(startPage, True)
		End Sub

		Private Sub richEditControl1_HyperlinkClick(ByVal sender As Object, ByVal e As HyperlinkClickEventArgs) Handles richEditControl1.HyperlinkClick
			BeginInvoke(CType(Function() AnonymousMethod1(e), MethodInvoker))
			e.Handled = True
		End Sub
		
		Private Function AnonymousMethod1(ByVal e As HyperlinkClickEventArgs) As Boolean
			NavigateTo(e.Hyperlink.NavigateUri, True)
			Return True
		End Function

		Private Sub richEditControl1_KeyDown(ByVal sender As Object, ByVal e As KeyEventArgs) Handles richEditControl1.KeyDown
			If currentHistoryItemIndex <> 0 AndAlso e.KeyCode = Keys.Back Then
				btnBackward_Click(btnBackward, EventArgs.Empty)
			End If
		End Sub

		Private Sub btnBackward_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles btnBackward.Click
			currentHistoryItemIndex -= 1
			NavigateTo(historyItems(currentHistoryItemIndex), False)
		End Sub

		Private Sub btnForward_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles btnForward.Click
			currentHistoryItemIndex += 1
			NavigateTo(historyItems(currentHistoryItemIndex), False)
		End Sub

		Private Sub NavigateTo(ByVal page As String, ByVal addToHistory As Boolean)
			If addToHistory Then
				Do While historyItems.Count > currentHistoryItemIndex + 1
					historyItems.RemoveAt(historyItems.Count - 1)
				Loop

				historyItems.Add(page)
				currentHistoryItemIndex = historyItems.Count - 1
			End If
			richEditControl1.LoadDocument(Path.Combine(basePath, page))
			UpdateUI()
		End Sub

		Private Sub UpdateUI()
			btnBackward.Enabled = currentHistoryItemIndex <> 0
			btnForward.Enabled = currentHistoryItemIndex <> historyItems.Count - 1
		End Sub
	End Class
End Namespace